### PR TITLE
Fix qrcode bugs

### LINKED ISSUE #22892: Segfault in QRCodeDetector.detectAndDecodeCurved
### System Information

OpenCV version: 4.2.0+dfsg-5 from Ubuntu 20.04 and 4.5.4 from Nix latest.
Operating System: Ubuntu 20.04 on AMD Ryzen 7 5700G
Python version: 3.8.10 from Ubuntu 20.04 and 3.10.8 from Nix latest.

### Detailed description

We get a segmentation fault calling QRCodeDetector.detectAndDecodeCurved for some images:

```
$ PYTHONFAULTHANDLER=1 python3 crash_repo.py
Fatal Python error: Segmentation fault

Current thread 0x00007fd41ab83740 (most recent call first):
  File "/xxx/crash_repo.py", line 5 in <module>

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, cv2 (total: 14)
Segmentation fault (core dumped)
```

This happens infrequently but with a given image it is 100% reproducible.

Seems similar to #19626 but they want a new issue over there.

### Steps to reproduce

Run the following minimal code:

```
import cv2 as cv

img = cv.imread("last_qr.png")
qcd = cv.QRCodeDetector()
qcd.detectAndDecodeCurved(img)
```

This file triggers it: last_qr

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### LINKED ISSUE #24011: QRCodeDetector - Seed point is outside of image in function 'cv::floodFill'
### System Information

OpenCV python version: 4.8.0.74
Operating System: Windows 10 22H1
Python version: 3.11.4

### Detailed description

QRCodeDetector `detectAndDecodeMulti` fails when there is a transition with QR codes.
I would have expected a "not detected" or "unable to decode" rather than an exception.

A similar issue was reported and fixed for v4.8.0. #21532 

```
floodfill.cpp:522: error: (-211:One of the arguments' values is out of range) Seed point is outside of image in function 'cv::floodFill'
```

### Steps to reproduce

floodfillerror_813

Run QRCodeDectector.detectAndDecodeMulti() with this image.

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### LINKED ISSUE #24450: Single black pixel breaks QR code detect
### System Information

OpenCV: a3b3a589f988ad6590999130d3b0d506dbb0844f

### Detailed description

Using an image [qrcode-kanji_mod.png](https://github.com/opencv/opencv/assets/25801568/781187fc-16b4-40bf-b4c2-aa65e470602c) with a single extra black pixel on specific location breaks detection.

### Steps to reproduce

```python
import cv2 as cv
img = cv.imread("qrcode-kanji_mod.png")
detect = cv.QRCodeDetector()
det = detect.detect(img)
print("1:", det)

img = cv.imread("qrcode-kanji_mod.png")
img[-1, 296] = 0
detect = cv.QRCodeDetector()
det = detect.detect(img)
print("2:", det)
```
```
1: (True, array([[[  9.,   4.],
        [632.,   4.],
        [632., 628.],
        [  9., 628.]]], dtype=float32))
2: (False, None)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by akretz**
This PR fixes #22892, #24011 and #24450 and adds regression tests using the images provided. I've also verified with the [benchmark](https://github.com/opencv/opencv_benchmarks/tree/develop/python_benchmarks/qr_codes) that this doesn't break anything there.

resolves #22892
resolves #24011
resolves #24450
Replaces https://github.com/opencv/opencv/pull/23802

Requires extra: https://github.com/opencv/opencv_extra/pull/1148

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
No regressions in QR code benchmark: https://github.com/opencv/opencv_benchmarks/pull/9

**Comment by dkurt** [comment]
There is one optional comment to test data: https://github.com/opencv/opencv_extra/pull/1148/files#r1490488836. If it's possible to test regression on grayscale image, I recommend to use it.

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -466,16 +466,20 @@ bool QRDetect::localization()
...
+        if (!list_lines_y.empty())
+        {
+            kmeans(list_lines_y, 3, labels,
+                TermCriteria( TermCriteria::EPS + TermCriteria::COUNT, 10, 0.1),
+                3, KMEANS_PP_CENTERS, localization_points);
 
-    fixationPoints(localization_points);
+            fixationPoints(localization_points);
+        }
+    }
```
Comment (asmorkalov) [inline modules/objdetect/src/qrcode.cpp:482]: `localization_points` should be reset in else branches, if `kmeans`/`fixationPoints` do not happen.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by AleksandrPanov — APPROVED** [review]
👍

**Comment by jabdoa2** [comment]
Thank you that :-)!

**Pull request merged**